File: mcp_gateway/mod_client.py
#!/usr/bin/env python3
"""
mod_client — HTTP client for the WitchModMCP in-game JSON-RPC server.

All HTTP communication with the game mod flows through this module.
Every request creates a fresh connection (thread-safe).
"""

import json
import os
import sys
import threading
import http.client

DEFAULT_MOD_PORT = 3100


def log(msg: str):
    """Log diagnostic messages to stderr (never stdout)."""
    print(f"[mod_client] {msg}", file=sys.stderr, flush=True)


class ModConnection:
    """Manages HTTP communication with the game mod's built-in JSON-RPC server.

    Connection-per-request design ensures thread safety between the
    heartbeat thread and MCP handler thread.
    """

    def __init__(self, port: int):
        self.port = port
        self._id_counter = 0
        self._id_lock = threading.Lock()

    # ── low-level HTTP helpers ───────────────────────────────────────

    def _request(self, method: str, path: str,
                 body: str | None = None,
                 timeout: int = 5) -> tuple[int, str]:
        """Send an HTTP request and return (status_code, response_body).

        Creates a fresh connection per call for thread safety.
        """
        conn = http.client.HTTPConnection("localhost", self.port, timeout=timeout)
        try:
            headers = {"Content-Type": "application/json"}
            conn.request(method, path, body, headers)
            resp = conn.getresponse()
            data = resp.read().decode("utf-8", errors="replace")
            return resp.status, data
        finally:
            conn.close()

    # ── public API ───────────────────────────────────────────────────

    def call_tool(self, method: str, params: dict | None = None) -> dict:
        """POST JSON-RPC to the mod. Normalises PascalCase keys to camelCase.

        Args:
            method: Tool name (e.g. "get_game_data", "eval_command").
            params: Tool arguments dict.

        Returns:
            Normalised JSON-RPC response dict (result / error / jsonrpc / id).

        Note:
            Tool calls use a 120s timeout because some C# tools (decompile_source,
            get_fight_state during complex fights, etc.) may take many seconds.
        """
        with self._id_lock:
            self._id_counter += 1
            req_id = self._id_counter
        req_body = json.dumps({
            "jsonrpc": "2.0",
            "id": req_id,
            "method": method,
            "params": params or {},
        })
        try:
            status, body = self._request("POST", "/", req_body, timeout=120)
            if status != 200:
                return {
                    "jsonrpc": "2.0",
                    "error": {
                        "code": -32000,
                        "message": f"Mod returned HTTP {status}: {body[:200]}",
                    },
                }
            data = json.loads(body)
            return self._lower_keys(data)
        except json.JSONDecodeError:
            return {
                "jsonrpc": "2.0",
                "error": {"code": -32700, "message": "Invalid JSON response from mod"},
            }
        except Exception as e:
            return {
                "jsonrpc": "2.0",
                "error": {"code": -32000, "message": f"Mod connection failed: {e}"},
            }

    def send_heartbeat(self, workspace_dir: str) -> tuple[bool, dict | None]:
        """POST /heartbeat — send a heartbeat to the mod (no auth).

        Returns:
            (ok, response_dict) where ok is True on status==200 and
            the response dict contains the parsed JSON body.
        """
        body = json.dumps({
            "workspacePath": workspace_dir,
            "pid": os.getpid(),
            "keepalive": True,
        })
        try:
            status, raw = self._request("POST", "/heartbeat", body)
            data = json.loads(raw)
            if status == 200 and data.get("status") == "ok":
                return True, data
            return False, data
        except json.JSONDecodeError:
            return False, {"error": "Invalid JSON response"}
        except Exception as e:
            return False, {"error": str(e)}

    # ── helpers ──────────────────────────────────────────────────────

    @staticmethod
    def _lower_keys(d):
        """Recursively convert dict keys from PascalCase to camelCase
        (handles C# Newtonsoft serialization: JsonRpc -> jsonRpc)."""
        if isinstance(d, dict):
            result = {}
            for k, v in d.items():
                key = k[0].lower() + k[1:] if k else k
                result[key] = ModConnection._lower_keys(v)
            return result
        elif isinstance(d, list):
            return [ModConnection._lower_keys(v) for v in d]
        return d

File: gateway/src/modClient.ts
/**
 * modClient.ts — HTTP client for the WitchModMCP in-game JSON-RPC server.
 *
 * Port of `mcp_gateway/mod_client.py`. Communication with the C# mod goes
 * through this module. Node uses the global `fetch` (no extra dependency).
 */
import { lowerFirstKey } from "./tools/camelCase.js";

export const DEFAULT_MOD_PORT = 3100;

// Long guard for unusual mod responses; all current C# tools are asynchronous
// and return quickly (see DecompileSourceTool etc.), so the old Python 120s
// default is intentionally NOT reproduced. Heartbeat uses a smaller timeout.
const TOOL_TIMEOUT_MS = 30_000;
const HEARTBEAT_TIMEOUT_MS = 5_000;

export interface ToolLogFn {
  (msg: string): void;
}

/** @param msg log to stderr (never stdout — stdout is the MCP channel). */
export function log(msg: string): void {
  console.error(`[mod_client] ${msg}`);
}

export interface ModToolResponse {
  jsonrpc?: string;
  id?: number | string;
  result?: unknown;
  error?: { code: number; message: string };
}

export class ModConnection {
  readonly port: number;
  private _idCounter = 0;

  constructor(port: number) {
    this.port = port;
  }

  private nextId(): number {
    return ++this._idCounter;
  }

  private async request(
    method: string,
    path: string,
    body: string | null,
    timeoutMs: number,
  ): Promise<[number, string]> {
    const url = `http://localhost:${this.port}${path}`;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    try {
      const resp = await fetch(url, {
        method,
        headers: body !== null ? { "Content-Type": "application/json" } : undefined,
        body: body !== null ? body : undefined,
        signal: controller.signal,
      });
      const text = await resp.text();
      return [resp.status, text];
    } finally {
      clearTimeout(timer);
    }
  }

  /**
   * POST a JSON-RPC tool call to the mod.
   * Returns the normalised JSON-RPC response dict. Keys are normalised from
   * PascalCase (C# Newtonsoft) to camelCase.
   */
  async callTool(method: string, params?: Record<string, unknown> | null): Promise<ModToolResponse> {
    const reqId = this.nextId();
    const body = JSON.stringify({
      jsonrpc: "2.0",
      id: reqId,
      method,
      params: params ?? {},
    });
    try {
      const [status, raw] = await this.request("POST", "/", body, TOOL_TIMEOUT_MS);
      if (status !== 200) {
        return {
          jsonrpc: "2.0",
          error: {
            code: -32000,
            message: `Mod returned HTTP ${status}: ${raw.slice(0, 200)}`,
          },
        };
      }
      const data = JSON.parse(raw);
      return lowerFirstKey(data) as ModToolResponse;
    } catch (err) {
      return {
        jsonrpc: "2.0",
        error: {
          code: err instanceof SyntaxError ? -32700 : -32000,
          message:
            err instanceof SyntaxError
              ? "Invalid JSON response from mod"
              : `Mod connection failed: ${(err as Error).message}`,
        },
      };
    }
  }

  /**
   * POST /heartbeat — send a heartbeat to the mod (no auth).
   * Returns [ok, responseDict]; ok is true when status==200 and status=="ok".
   */
  async sendHeartbeat(workspaceDir: string): Promise<[boolean, Record<string, unknown>]> {
    const body = JSON.stringify({
      workspacePath: workspaceDir,
      pid: process.pid,
      keepalive: true,
    });
    try {
      const [status, raw] = await this.request("POST", "/heartbeat", body, HEARTBEAT_TIMEOUT_MS);
      const data = JSON.parse(raw);
      const ok = status === 200 && (data as { status?: string }).status === "ok";
      return [ok, data];
    } catch (err) {
      return [false, { error: (err as Error).message }];
    }
  }
}
